core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+			pmmr::insertion_to_pmmr_index(leaf_offset + segment_size)
+				+ (self.identifier.log_size as u64)
+		} else {
+			last_pos
+		};
+
+		(first, last)
+	}
+
+	fn get_hash(&self, pos: u64) -> Result<Hash, SegmentError> {
+		self.hashes
+			.binary_search_by(|&(p, _)| p.cmp(&pos))
+			.map(|p| self.hashes[p].1)
+			.map_err(|_| SegmentError::MissingHash(pos))
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: Readable + Writeable + Debug,
+{
+	/// Generate a segment from a PMMR
+	pub fn from_pmmr<U, B>(
+		segment_id: SegmentIdentifier,
+		pmmr: &ReadonlyPMMR<'_, U, B>,
+		prunable: bool,
+	) -> Result<Self, SegmentError>
+	where
+		U: PMMRable<E = T>,
+		B: Backend<U>,
+	{
+		let mut segment = Segment {
+			identifier: segment_id,
+			hashes: Vec::new(),
+			leaf_data: Vec::new(),
+			proof: SegmentProof::empty(),
+		};
+
+		let last_pos = pmmr.unpruned_size();
+		if segment.segment_unpruned_size(last_pos) == 0 {
+			return Err(SegmentError::NonExistent);
+		}
+
+		// Fill leaf data and hashes
+		let (segment_first_pos, segment_last_pos) = segment.segment_pos_range(last_pos);
+		for pos in segment_first_pos..=segment_last_pos {
+			if pmmr::is_leaf(pos) {
+				if let Some(data) = pmmr.get_data_from_file(pos) {
+					segment.leaf_data.push((pos, data));
+					continue;
+				} else if !prunable {
+					return Err(SegmentError::MissingLeaf(pos));
+				}
+			}
+			// TODO: optimize, no need to send every intermediary hash
+			if prunable {
+				if let Some(hash) = pmmr.get_from_file(pos) {
+					segment.hashes.push((pos, hash));
+				}
+			}
+		}
+
+		if segment.leaf_data.is_empty() {
+			return Err(SegmentError::Pruned);
+		}
+
+		// Segment merkle proof
+		segment.proof =
+			SegmentProof::generate(pmmr, last_pos, segment_first_pos, segment_last_pos)?;
+
+		Ok(segment)
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: PMMRIndexHashable,
+{
+	/// Calculate root hash of this segment
+	pub fn root(&self, last_pos: u64, bitmap: Option<&Bitmap>) -> Result<Hash, SegmentError> {
+		let (segment_first_pos, segment_last_pos) = self.segment_pos_range(last_pos);
+		let mut hashes =
+			Vec::<Option<Hash>>::with_capacity(2 * (self.identifier.log_size as usize));
+		let mut leaves = self.leaf_data.iter();
+		for pos in segment_first_pos..=segment_last_pos {
+			let height = pmmr::bintree_postorder_height(pos);
+			let hash = if height == 0 {
+				// Leaf
+				if bitmap
+					.map(|b| {
+						let idx_1 = pmmr::n_leaves(pos);
+						let idx_2 = if pmmr::is_left_sibling(pos) {
+							idx_1 + 1
+						} else {
+							idx_1 - 1
+						};
+						b.contains(idx_1 as u32) || b.contains(idx_2 as u32)
+					})
+					.unwrap_or(true)
+				{
+					// We require the data of this leaf if either the mmr is not prunable or if
+					// the bitmap indicates it (or its sibling) should be here
+					// TODO: possibly remove requirement on the sibling when we no longer support
+					//  syncing through the txhashset.zip method
+					let data = leaves
+						.find(|&&(p, _)| p == pos)
```

Comment:
Is this not always the next element in leaf_data? Do we need to find it here?
i.e. We just iterate over data as the leaves are already sorted?

Reply:
Similar to how we pop() twice for the left and right hashes below. We can do something similar here for left and right leaves (at least currently where we know both will exist due to pruning impl).

Reply:
The leaves are guaranteed to be sorted, yes. However, my understanding is that if two sibling leaves are spent, but compaction has not been run, they still exist in the backend. During segment construction we include any leaf that still exists in the backend, without consulting a bitmap. That means the next leaf is not always the leaf we expect.

Reply:
Ok yeah this makes sense. So we are leveraging the iterator and just need to iterate once, but here we may skip over redundant leaves. :+1: